### PR TITLE
Configurable temporary directory

### LINKED ISSUE #12910: Move temp dir to /data
Based on: https://github.com/owncloud/core/issues/5949#issuecomment-67312666

**Background:**
Most of the external storage types do right now not support streaming. This means, that any file transferred needs to be buffered at a temporary space on the server who runs owncloud. This can cause severe space issues, even if the space needed is only for a limited period of time. Right now, the temporary location is defined at https://github.com/owncloud/core/blob/master/lib/base.php#L980
It is either getting the windows version of the tmp dir or via php (sys_get_temp_dir) where unix will put it to.

**Suggestion:**
Define the temp dir used at /data (in eg a subdirectory).

The advantage having the temp dir in /data would be, that the temp dir is independent of the OS and webserver used and "outsourceable". 'datadirectory' in config.php is a configurable item...

We would immediate avoid possible space problems at the server caused by temporary files and owncloud related things stay together and are not distributed around.

If the /data directory stays where it is originally setup, nothing will happen negatively as it uses the same physical location of the owncloud installation (and where eg the php temp dir ist set), but as mentioned above it could be relocated easily.

### LINKED ISSUE #14886: opendir(/tmp): failed to open dir: Permission denied at /.../lib/private/tempmanager.php#134
Version: 8.0.2 and 7.0.5

With Version 8.0.2 and 7.0.5 my log is full of those messages:
opendir(/tmp): failed to open dir: Permission denied at /.../lib/private/tempmanager.php#134 

/tmp has the following permissions:
drwxrwxrwt  4 root         root     4096 Mär 13 18:14 tmp

Those look fine.

This owncloud instance is hosted by all-inkl, might this be a php configuration issue?

### PR DESCRIPTION
**Pull request by RobinMcCorkell**
This allows the temporary directory used by ownCloud to be configured in `config.php`.

Replaces #16636 

Fixes #14886, fixes #12910

cc @mmattel @MorrisJobke @PVince81 @DeepDiver1975

### CODE REVIEW

File: lib/base.php [hunk 1]
```diff
@@ -1108,27 +1108,16 @@ protected static function tryFormLogin() {
 		}
 		return true;
 	}
-}
 
-if (!function_exists('get_temp_dir')) {
```
Comment (MorrisJobke) [inline lib/base.php]: Why? I guess this can cause problems or wasn't there a reason to inclide this check?

Comment (MorrisJobke) [inline lib/base.php]: ah ... you added it as an `\OC` method :)

File: tests/lib/tempmanager.php [hunk 2]
```diff
@@ -35,18 +36,27 @@ protected function setUp() {
...
 	 * @return \OC\TempManager
 	 */
-	protected function getManager($logger = null) {
+	protected function getManager($logger = null, $config = null) {
 		if (!$logger) {
 			$logger = new NullLogger();
 		}
-		return new \OC\TempManager($this->baseDir, $logger);
+		if (!$config) {
+			$config = \OC::$server->getConfig();
```
Comment (MorrisJobke) [inline tests/lib/tempmanager.php]: why not mock it?

File: lib/private/tempmanager.php [hunk 3]
```diff
@@ -190,4 +195,67 @@ protected function getOldFiles() {
...
+	 * @return string Path to the temporary directory or null
+	 * @throws \UnexpectedValueException
+	 */
+	public function getTempBaseDir() {
+		if ($this->tmpBaseDir) {
+			return $this->tmpBaseDir;
+		}
+
+		$directories = [];
+		if ($temp = $this->config->getSystemValue('tempdirectory', null)) $directories[] = $temp;
```
Comment (MorrisJobke) [inline lib/private/tempmanager.php]: This breaks our coding guidelines

**Comment by karlitschek** [comment]
:+1: from me beside the mentioned issues.

File: tests/lib/tempmanager.php (hunk 2 above)
Comment (Xenopathic) [inline tests/lib/tempmanager.php]: I was lazy :smile:

File: lib/private/tempmanager.php (hunk 3 above)
Comment (Xenopathic) [inline lib/private/tempmanager.php]: This was originally written by @mmattel, and in the previous PR there were no comments about it. For so many ifs right next to each other, is it really necessary to have curly braces? :cry:

File: tests/lib/tempmanager.php (hunk 2 above)
Comment (MorrisJobke) [inline tests/lib/tempmanager.php]: > I was lazy :smile:

Don't be lazy and fix it :smiley:

File: lib/private/tempmanager.php (hunk 3 above)
Comment (MorrisJobke) [inline lib/private/tempmanager.php]: Yep. You know what happened once a while back with apple and the goto commands :wink:

Comment (Xenopathic) [inline lib/private/tempmanager.php]: Good point

**Comment by Xenopathic** [comment]
@owncloud-bot retest this please

**Comment by karlitschek** [comment]
@Xenopathic where are we with that? Would be nice to have it in 8.2

**Comment by Xenopathic** [comment]
Waiting for reviewers, cc @MorrisJobke

File: lib/public/itempmanager.php [hunk 4]
```diff
@@ -58,4 +58,12 @@ public function clean();
...
 	 */
 	public function cleanOld();
+
+	/**
+	 * Get the temporary base directory
+	 *
+	 * @return string
+	 * @since 8.2.0
+	 */
+	public function getTempBaseDir();
```
Comment (MorrisJobke) [inline lib/public/itempmanager.php]: @icewind1991 Isn't there already a public method to get a temp directory? What is the difference here?

Comment (icewind1991) [inline lib/public/itempmanager.php]: FFrom what I can tell this is the base folder we create the other folders in, I dont know if we need to expose this as public

**Comment by rullzer** [comment]
Looks good. And a very useful addition.
Please add it also to the config.sample.php.

**Comment by mmattel** [comment]
I have written a documentation some months ago: https://github.com/owncloud/documentation/pull/779
But this one got accidatially merged and reverted: https://github.com/owncloud/documentation/pull/1254
Maybe the revert can be reverted to get it implemented...
In my original PR core https://github.com/owncloud/core/pull/16636, I have written a config.php.sampe text which also could be reused if wanted.

File: lib/public/itempmanager.php (hunk 4 above)
Comment (Xenopathic) [inline lib/public/itempmanager.php]: I added the public API since the original `get_temp_dir()` was accessible.

**Comment by Xenopathic** [comment]
@rullzer @mmattel Please review

**Comment by rullzer** [comment]
Yep thanks for the example config.
:+1:

**Comment by karlitschek** [comment]
:+1:

File: lib/base.php [hunk 5]
```diff
@@ -1108,27 +1108,16 @@ protected static function tryFormLogin() {
...
-		if ($temp = getenv('TMPDIR')) return $temp;
-		$temp = tempnam(__FILE__, '');
-		if (file_exists($temp)) {
-			unlink($temp);
-			return dirname($temp);
-		}
-		if ($temp = sys_get_temp_dir()) return $temp;
-
-		return null;
+		return \OC::$server->getTempManager()->t_get_temp_dir();
```
Comment (DeepDiver1975) [inline lib/base.php]: t_get_temp_dir ?????? not know 

bildschirmfoto von 2015-09-14 18-07-42

**Pull request merged**